### ISSUE TITLE
Currently no error when poping a Required key from a TypedDict & still return correct type.

### ISSUE THREAD

**Issue by Daraan** [microsoft/pyright#10087]
Consider the following [example (pyright playground)](https://pyright-play.net/?enableExperimentalFeatures=true&code=GYJw9gtgBALgngBwJYDsDmB9ApgDxllAZyTCKiQgTBBigBVEsATAESQGMYAaKAJSwCOAVyQhmPAHJgY-YaPFQAgijg8GCLIoA2SAIaEAUAfZb9hJQAp1zNpx4xpurQF46IIVgCUALgNR-UMBgYN6BWmC6MH4BQWAATKFSMoIiYkwA2sDhkQC6RkxYwFBoWDAYihaeUAC0AHxKoQB0zQa6UM7FpeWVRmIAblhOGPAaFrqNVAgWAESEkFgA1lhw055VUADEUEIoCyhgAO4oUEtwRlsAoiDgIKEAwmBCWkxQ%2B7RmSGjHSPjQqPSMVgcKJbKDTMAAIwAVlhONNyOY3lAPl9dBCtFhYGAwVkIjBpq10tNYnFpjl2siJmApsTgqsAgytmI5GkTssgA)

```python
from typing_extensions import TypedDict, Required, NotRequired, Any, TypeAlias

class A(TypedDict, total=True):
    foo: float
    foo2: NotRequired[float]

def get_A() -> A: ...
a = get_A()

reveal_type(a.pop("somekey"))  # unknown key -> object # OK

# Error: Could not assign item in TypedDict
#  "object" is not assignable to "float"
a["foo2"] = a.pop("foo")  # pop required key; warning only on the left hand sight.
```

This rises an unexpected error that `object` is not assignable to the `foo2` key. I expected it to be the same type `float`.
Of course this violates the structure of the `TypedDict`, the currently inferred `pop` only works for not-required keys to respect that, I assume.

**Describe the solution you’d like**

1) Poping a required key should return the correct type, `float` in this case
2) As it violates the structure raise error like in the `del a["foo"] case.

**Comment by erictraut** [comment]
When pyright synthesizes the methods of a TypedDict, it creates an overload for `pop`. In particular, it creates one overload signature for each key that is both not required and not read-only. It also provides a catch-all overload signature that accepts any key and returns an `object`. This overload provides for dynamic string values

 While I could theoretically omit that final catch-all overload, it was requested by numerous pyright users, and I think I'd get a lot of pushback if I deleted it. It comes down to a question of which is better — false positives in the dynamic case or false negatives in the case that you've highlighted above. I don't think there's a "correct" answer here, but based on feedback from pyright users, I'm going to leave the current functionality in place.

**Comment by Daraan** [comment]
Thank you once again for your reply, that catch-all, support for dynamic keys should definitely stay in place.

As a follow up, I am not sure, does your response also address the second point I mentioned. For example the following case does not raise any error, but would cause a runtime error as I deleted a required key via `pop`.

```python
a.pop("foo")       # poping required keys is fine.
...
foo = a["foo"]    # OK
```

**Comment by erictraut** [comment]
As I said, the synthesized `pop` overload includes an overload for each of the not-required fields. It also includes a catch-all that accepts any `str` value to support dynamic values. With this catch-all overload in place, a call like `a.pop("foo")` will not be flagged as an error because it is covered by the catch-all.
